untidy/contaminators.py

```diff
@@ -3,129 +3,197 @@
+        data to be contaminated with superfluous characters
+    corruption_level: int, optional
         level of corruption, should be between 0 and 10, where 0 leaves the dataset as is, 10
         is the highest level of contamination
 
     Returns
     -------
-    list of indeces to be contaminated
+    data: pd.DataFrame or pd.Series
+        contaminated dataset
     """
+    data = clean_data.copy()
 
-    # define amount of missing values to introduce
-    contamination_fraction = round(num_rows * num_cols * corruption_level/10 * 0.6)
+    # Find data cells to contaminate
+    idxs_to_contaminate = get_random_indices(
+        data, col_type="str", corruption_level=corruption_level
+    )
 
-    # generate randomized missing values indeces
-    
-    if num_cols>1:
-        cont_idxs = zip(random.choices(range(num_rows), k=contamination_fraction),
-                    random.choices(range(num_cols), k=contamination_fraction))
-    else:
-        cont_idxs = zip(random.choices(range(num_rows), k=contamination_fraction),
-                    [0 for i in range(contamination_fraction)])
-
-    return cont_idxs
+    # Perform contamination
+    noise_chars = "%&$?!#"
+    whitespace = "  "
 
+    for idx in idxs_to_contaminate:
+        # add either extra whitespace or a superfluous character
+        noise_char = random.choice(noise_chars)
+        noise = random.choice([noise_char, whitespace])
+        data.iloc[idx] = str(data.iloc[idx]).replace("nan", "") + noise
 
+    return data
 
-""" Functions to contaminate any columns: """
 
-def add_nans(clean_data, corruption_level = 4):
+def change_str_encoding(clean_data, corruption_level=4):
     """
-    Introduce missing values in clean data
+    Changes the string encoding of text data.
 
     Parameters
     ----------
     clean_data: pd.Series or pd.DataFrame
-        data to be contaminated with missing values
+        data to be contaminated with superfluous characters
     corruption_level: int, optional
         level of corruption, should be between 0 and 10, where 0 leaves the dataset as is, 10
         is the highest level of contamination
 
     Returns
     -------
-    contaminated data as pd.DataFrame or pd.Series
-
+    data: pd.DataFrame or pd.Series
+        contaminated dataset
     """
     data = clean_data.copy()
-    print(data)
-    
-    num_rows, num_cols = get_data_dims(data)
 
-    nan_idxs = get_random_indices(num_rows, num_cols, corruption_level = corruption_level)
-    
-    # insert missing values
-    for x, y in nan_idxs:
-        if isinstance(data, pd.DataFrame):
-            data.iloc[x:x+1, y:y+1] = np.nan
+    if isinstance(data, pd.DataFrame):
+        # Find random columns to contaminate
+        str_cols = list(data.select_dtypes(include=["object"]).columns)
+        num_str_cols = len(str_cols)
+        possible_num_cols_to_contaminate = [
+            int(np.ceil(n)) for n in np.linspace(0, int(num_str_cols) / 2, 11)
```

Comment:
Doesn't get_random_indices do the same job of getting the possible columns to contaminate?

Reply:
get_random_indices selects data cells, this one selects columns